Sources/ServiceLifecycle/ServiceGroup.swift

```diff
@@ -777,6 +850,45 @@ public actor ServiceGroup: Sendable, Service {
             cancellationTimeoutTask = nil
         }
     }
+
+    private func addServiceTask(
+        group: inout ThrowingTaskGroup<ChildTaskResult, Error>,
+        service serviceConfiguration: ServiceGroupConfiguration.ServiceConfiguration,
+        gracefulShutdownManager: GracefulShutdownManager,
+        index: Int
+    ) {
+        // This must be addTask and not addTaskUnlessCancelled
+        // because we must run all the services for the shutdown logic to work.
+        group.addTask {
+            return await TaskLocals.$gracefulShutdownManager.withValue(gracefulShutdownManager) {
+                do {
+                    try await serviceConfiguration.service.run()
+                    return .serviceFinished(service: serviceConfiguration, index: index)
+                } catch {
+                    return .serviceThrew(service: serviceConfiguration, index: index, error: error)
+                }
+            }
+        }
+    }
+
+    private func addAddedServiceListenerTask(
+        group: inout ThrowingTaskGroup<ChildTaskResult, Error>,
+        channel: AsyncChannel<ServiceGroupConfiguration.ServiceConfiguration>
+    ) {
+        group.addTask {
+            return await withTaskCancellationHandler {
+                var iterator = channel.makeAsyncIterator()
+                if let addedService = await iterator.next() {
+                    return .newServiceAdded(addedService)
+                }
+
+                return .gracefulShutdownFinished
+            } onCancel: {
+                // Without this we can get stuck in `addService` if the group
```

Comment:
Can we change this comment to reflect our discussion above and just say `AsyncChannel` is MPSC hence we need to shut it down on cancel.